src/passes/MemoryPacking.cpp

```diff
@@ -23,141 +23,394 @@
+      bool isFunctionParallel() override { return false; }
+
+      Replacements& replacements;
+
+      Replacer(Replacements& replacements) : replacements(replacements){};
+
+      void visitMemoryInit(MemoryInit* curr) {
+        auto replacement = replacements.find(curr);
+        assert(replacement != replacements.end());
+
+        auto segmentIndex = replacement->second.first;
+        auto& ranges = replacement->second.second;
+        assert(ranges.size() > 0);
+
+        // If there was no transformation, only update the index
+        if (ranges.size() == 1 && !ranges.front().isZero) {
+          curr->segment = segmentIndex;
+          return;
+        }
+
+        Builder builder(*getModule());
+
+        // Nonconstant offsets or sizes will have inhibited splitting
+        size_t start = curr->offset->cast<Const>()->value.geti32();
+        size_t end = start + curr->size->cast<Const>()->value.geti32();
+        size_t bytesWritten = 0;
+
+        // Split curr into multiple memory.inits and memory.fills, storing the
+        // original base destination in a local if it is not a constant
+        Expression* result = nullptr;
+        Index destVar = -1;
+        if (!curr->dest->is<Const>()) {
+          destVar = builder.addVar(getFunction(), i32);
+          result = builder.makeLocalSet(destVar, curr->dest);
+        }
+        for (auto& range : ranges) {
+          // Skip ranges that do no intersect the range of the memory.fill
+          if (range.end <= start) {
+            if (!range.isZero) {
+              segmentIndex++;
+            }
+            continue;
+          }
+          if (range.start >= end) {
+            break;
+          }
+
+          Expression* dest;
+          if (auto* c = curr->dest->dynCast<Const>()) {
+            dest = builder.makeConst(
+              Literal(int32_t(c->value.geti32() + bytesWritten)));
+          } else {
+            dest = builder.makeLocalGet(destVar, i32);
+            if (bytesWritten > 0) {
+              Const* addend = builder.makeConst(Literal(int32_t(bytesWritten)));
+              dest = builder.makeBinary(AddInt32, dest, addend);
+            }
+          }
+
+          size_t bytes =
+            std::min(range.end, end) - std::max(range.start, start);
+          bytesWritten += bytes;
+          Expression* size = builder.makeConst(Literal(int32_t(bytes)));
+
+          if (range.isZero) {
+            Expression* value = builder.makeConst(Literal::makeZero(i32));
+            result = builder.blockify(
+              result, builder.makeMemoryFill(dest, value, size));
+          } else {
+            size_t offsetBytes = std::max(start, range.start) - range.start;
+            Expression* offset =
+              builder.makeConst(Literal(int32_t(offsetBytes)));
+            result = builder.blockify(
+              result, builder.makeMemoryInit(segmentIndex, dest, offset, size));
+            segmentIndex++;
+          }
+        }
+        if (result) {
```

Comment:
Can `result` be nullptr at this point, unless `ranges.size() == 0`, given that we make `memory.fill` or `memory.init` within the loop? And we have `assert(ranges.size() > 0)` above.

Reply:
Yes, a zero-length memory.init will not intersect any ranges so the loop will not add any instructions. I've added new tests for this.